### ISSUE TITLE
Require every check to have a category

### ISSUE THREAD

**Issue by felixarntz** [10up/plugin-check#203]
### Is your enhancement related to a problem? Please describe.

Since there are already at this point different kinds of checks available in the plugin checker, it's going to be crucial to be able to only run specific kinds of checks for specific use-cases. For example, the plugin review team may only be interested in plugin minimum requirement checks for the wordpress.org repository. On the flipside, someone building a plugin for a client may be interested in all checks except that one. There may be people that want to only run security checks. Etc

Right now, you can only run specific checks by passing the check slugs, however as the number of checks will grow further, this will soon become unreasonable as an approach (if it isn't already). We should therefore introduce the concept of **check categories**: Every check should have a category, from a group of predefined categories, and there should be a mechanism to only run checks of a certain category.

This issue should be focused only on implementing the basic infrastructure for categories:
* A place where all available categories are defined
* A modification to the check interface to ensure every check has a category attached
* A way to filter a list of checks to by certain categories.

A follow-up issue should then be responsible for offering a way to actually use this infrastructure via a new optional CLI argument and a new optional UI element in WP Admin.

### Designs

* Introduce a `Check_Categories` class.
* Add class constants (`const`) to the class for all available categories (see below).
* Implement a public static method `get_categories()` which simply returns the array of all categories.
* Implement a public static method `filter_checks_by_categories( array $checks, array $categories )` that filters a given list of check instances to only return those that have one of the provided categories.

For the way that checks declare their category, we should add a public method `get_category()` to the `Check` interface.
* Usage example: `$check->get_category()`
* Implementation example:
```php
class Code_Obfuscation_Check {
    public function get_category() {
        return Check_Categories::CATEGORY_PLUGIN_REPO;
    }
}
```

For the available categories, I think we can start with the following:
* `CATEGORY_GENERAL`
* `CATEGORY_PLUGIN_REPO`
* `CATEGORY_SECURITY`
* `CATEGORY_PERFORMANCE`
* `CATEGORY_ACCESSIBILITY`

### Alternatives considered

For checks to express their category, we could alternatively add a class constant like `CATEGORY` to the `Check interface and set it to some "default" category. Then every check has to define its own override of that class constant.
* Usage example: `Code_Obfuscation_Check::CATEGORY`
* Implementation example:
```php
class Code_Obfuscation_Check {
    const CATEGORY = Check_Categories::CATEGORY_PLUGIN_REPO;
}
```
The benefit would be a bit less code to write for every check (a constant is less than a method). However, this would not work for checks that use separate instances of the same class, which e.g. is relevant for #204. Therefore this is not a good option.

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by felixarntz** [comment]
@spacedmonkey @jjgrainger @mukeshpanchal27 Please review the above, let's discuss here how we want to proceed with this.

cc @eclarke1 @joemcgill

**Comment by mukeshpanchal27** [comment]
@felixarntz Ac's look good to me. 

One question: Do we need to use method `get_category` or constant a `CATEGORY` for category?

**Comment by jjgrainger** [comment]
Thanks @felixarntz 

AC's look good. I think going with a method `get_category` will provide more flexibility. To work with #204 would we also need to pass the category via the constructor?

**Comment by felixarntz** [comment]
@jjgrainger Maybe rather than using the constructor (which may get used more and more if we add features and thus get too complicated), we could implement a trait with the `get_category()` method and a `set_category( $category )` method. Then we could use that method on instances of the same class, using the trait.

**Comment by spacedmonkey** [comment]
I think checks should have the ability to be assigned to multiple categories. A check should for example be assigned to the base line check and a performance check. Or base line and translation. 

I think also the categories should be filterable. I would love if VIP GO or human made had they own checks and they could have they own rulesets, the same as PHPCS coding standard rule sets. 

I also wonder if we need to break up the `Plugin_Review_PHPCS_Check` check. So the plugin review checks, like database, php and security.

**Comment by felixarntz** [comment]
@spacedmonkey 
> I think checks should have the ability to be assigned to multiple categories. A check should for example be assigned to the base line check and a performance check. Or base line and translation.

Not sure about this. To me this feels like a theoretical idea, but when would it actually apply? The examples you made I'd say aren't really relevant, there are no performance checks that are a baseline requirement for plugins, and translations isn't a specific check category, I think that's too specific of a topic. I think we can always add this later if an actual need comes up for multiple categories on a single check.

> I think also the categories should be filterable. I would love if VIP GO or human made had they own checks and they could have they own rulesets, the same as PHPCS coding standard rule sets.

I like that idea. Maybe we can add a method like `register_category()` in the future. I'm always leaning towards starting with a baseline and expanding in the future as needed, otherwise we may end up features that nobody may ever use.

> I also wonder if we need to break up the `Plugin_Review_PHPCS_Check` check. So the plugin review checks, like database, php and security.

In principle I agree, though this comes from the plugin review team directly, so I'd prefer not to tamper with it until we have a closer collaboration with them and can get their thoughts on it.

**Comment by spacedmonkey** [comment]
I feel very strongly about this. Checks should have the ability to assigned to multiple category. I could see where a check could be for both, performance and security. Or security and translations. 

I think it would be a mistake not to add this functionality now, as I think it will be needed in the future.

**Comment by felixarntz** [comment]
@spacedmonkey Again, what would be an example check for this? I would like to see a realistic example this would apply to so that we can be sure it is a real problem to address.

**Comment by joemcgill** [comment]
I think I agree with @spacedmonkey here. Enforcing checks to a 1:1 taxonomy structure is likely to become too limiting. I can imagine teams wanting to run a set of checks based on a specific topic, like checks related to the editor, or a set of minimum checks that cover several different topics, like performance, a11y, security, etc. creating the possibility of a 1:n structure would be more flexible.

**Comment by felixarntz** [comment]
@joemcgill @spacedmonkey Okay, let's go with that then. In that case, we could probably change the interface for a check to have a `get_categories() : array` method instead of `get_category() : string`.

The only drawback of that is that the interface then technically allows to return an empty array which should never be the case, as we want every check to have at least one category. I guess that's okay. Since the method has to be implemented by every check, it should be intuitive enough to return at least one category. Additionally, we should clarify that in the method's doc block.

@mukeshpanchal27 Could you apply these changes to your PR #220 please?
